src/clj_irods/icat.clj

```diff
@@ -0,0 +1,207 @@
+    (->> [path ::paged-folder-listing zone user entity-type info-types sort-column sort-direction limit offset]
+         (cache/cached-or-nil (:cache irods)))))
+
+(defn maybe-paged-folder-listing
+  [& args]
+  (delay (deref (apply paged-folder-listing args))))
+
+(defn- resolve*
+  [depth nested-map]
+  (into {} (map
+             (fn [[k v]] [k (if (= depth 0) (if (delay? v) (cache/rethrow-if-error @v) v) (resolve* (dec depth) v))])
+             nested-map)))
+
+(defn all-cached-listings
+  [irods user zone path]
+  (let [all (get-in @(:cache irods) [path ::paged-folder-listing zone user])]
+    (when (seq all)
+    (delay
+      (resolve* 5 all)))))
+
+(defn- filter-listings*
+  ([filters nested-map]
+   (filter-listings* (count filters) filters nested-map))
+  ([depth filters nested-map]
+   (if (zero? depth)
+     nested-map
+     (let [this-filter (nth filters (- (count filters) depth))]
+       (into {}
+             (map (fn [[k v]] [k (filter-listings* (dec depth) filters v)])
+                  (if (nil? this-filter)
+                    nested-map
+                    (filter (fn [[k v]] (= k this-filter)) nested-map))))))))
+
+(defn filtered-cached-listings
+  [irods user zone path & {:keys [entity-type sort-column sort-direction limit offset info-types]}]
+  (let [all (all-cached-listings irods user zone path)]
+    (when all
+      (let [filters [entity-type info-types sort-column sort-direction limit offset]
+            cached  (filter-listings* filters @all)]
+        (when (seq cached)
+          (delay cached))))))
+
+(defn- keys-in [m]
+  (if (map? m)
+    (vec
+      (mapcat (fn [[k v]]
+                (let [sub (keys-in v)
+                      nested (map #(into [k] %) (filter (comp not empty?) sub))]
+                  (if (seq nested)
+                    nested
+                    [[k]])))
+              m))
+    []))
+
+;; NOTE: this merges only directly adjacent ranges -- overlapping ranges will remain separate. This could be less than ideal.
+;; e.g. if we have three ranges, (0 30) (10 40) (20 50), none of them will merge even though we have everything in (0 50)
+;; I'm hoping this doesn't matter much in practice, since it's usually not all that useful to grab overlapping ranges.
+(defn- merge-listing
+  [extant to-merge limit offset]
+  (let [ranges (into {} (mapcat (fn [[k v]] (mapv #(vector (+ % k) [k %]) (keys v))) extant))]
+    (if (contains? ranges offset) ;; is there an existing range that ends at our start point
+      (let [[old-limit old-offset] (get ranges offset) ;; there is, so extend it with these new values
+            extant-values (get-in extant [old-limit old-offset])
+            new-values    (concat extant-values to-merge)]
+        (into {} (remove (fn [[k v]] (empty? v))
+          (-> extant
+              (update-in [old-limit] dissoc old-offset) ;; effectively, dissoc-in [old-limit old-offset]
```

Comment:
There's no `dissoc-in` function in clojure. The first and third lines combine to mimic it -- the third dissoc's the `offset` out of the map by `limit`, and the first line then removes any `limit` ones that don't have any `offset` keys in them any more.

Reply:
`weavejester/medley` has a `dissoc-in` function if you'd like to refactor it.

Reply:
I'll definitely take a look at that, see if it does what I need, thanks! The less code I can accidentally introduce bugs in and all that.